lib/membrane/aac/parser.ex

```diff
@@ -38,42 +48,86 @@ defmodule Membrane.AAC.Parser do
+                  - `esds` - the field will be an `esds` MP4 atom. `avg_bit_rate` and `max_bit_rate` can be 
+                    additionally provided and will be included in the `esds`. If not known they should be set to 0.
+                  - `audio_specific_config` - the field will be a decoder configuration data AudioSpecificConfig(), 
+                as defined in ISO/IEC 14496-3.
                 """
               ]
 
   @type timestamp :: Ratio.t() | Membrane.Time.t()
 
-  @impl true
-  def handle_init(_ctx, options) do
-    {output_config, options} = Map.pop!(options, :output_config)
+  defmodule State do
+    @moduledoc false
+
+    @type t :: %__MODULE__{
+            samples_per_frame: AAC.samples_per_frame(),
+            out_encapsulation: AAC.encapsulation(),
+            audio_specific_config: binary(),
+            output_config: :audio_specific_config | :esds,
+            avg_bit_rate: non_neg_integer(),
+            max_bit_rate: non_neg_integer(),
+            leftover: binary(),
+            timestamp: Membrane.AAC.Parser.timestamp(),
+            in_encapsulation: AAC.encapsulation() | nil
+          }
+
+    @enforce_keys [
+      :samples_per_frame,
+      :out_encapsulation,
+      :audio_specific_config,
+      :output_config,
+      :avg_bit_rate,
+      :max_bit_rate
+    ]
+    defstruct @enforce_keys ++
+                [
+                  leftover: <<>>,
+                  timestamp: 0,
+                  in_encapsulation: nil
+                ]
+  end
 
+  @impl true
+  def handle_init(_ctx, opts) do
     {output_config, avg_bit_rate, max_bit_rate} =
-      case output_config do
+      case opts.output_config do
         {:esds, avg_bit_rate, max_bit_rate} -> {:esds, avg_bit_rate, max_bit_rate}
         config -> {config, 0, 0}
       end
 
-    state =
-      options
-      |> Map.from_struct()
-      |> Map.merge(%{
-        leftover: <<>>,
-        timestamp: 0,
-        in_encapsulation: nil,
-        output_config: output_config,
-        avg_bit_rate: avg_bit_rate,
-        max_bit_rate: max_bit_rate
-      })
+    state = %State{
+      samples_per_frame: opts.samples_per_frame,
+      out_encapsulation: opts.out_encapsulation,
+      audio_specific_config: opts.audio_specific_config,
+      output_config: output_config,
+      avg_bit_rate: avg_bit_rate,
+      max_bit_rate: max_bit_rate
+    }
 
     {[], state}
   end
 
   @impl true
   def handle_stream_format(:input, %AAC{} = stream_format, _ctx, state) do
-    stream_format = Config.parse_config(stream_format)
+    stream_format =
+      stream_format
+      |> Map.update!(:config, fn
+        nil when is_binary(state.audio_specific_config) ->
+          {:audio_specific_config, state.audio_specific_config}
+
+        nil when is_nil(state.audio_specific_config) ->
+          nil
+
+        non_nil_config ->
+          Membrane.Logger.warning(
+            "AAC config provided both with stream_format and options, ignoring the latter"
+          )
```

Comment:
what about a case where `stream_format.config` is not nil and `state.audio_specific_config` is nil? We shouldn't print that message then.

Reply:
I would also consider rewriting it with some `cond` clause as I think it could be more descriptive if you wrote something like: `config == nil and state.audio_specific_config == nil`